File: Src/ECScape.Core/Systems/PhysicsSystem.cs
using ECScape.Core.Components;
using ECScape.Core.Engine;
using ECScape.Core.Entities;

namespace ECScape.Core.Systems;

internal sealed class PhysicsSystem : ISystem
{
    public void Update(World world, float deltaTime)
    {
        Parallel.ForEach(world.Entities
            .Where(x => x.HasComponent<Position>() && x.HasComponent<AffectedByGravity>()), entity => Update(entity, world, deltaTime));
    }

    public static void Update(Entity entity, World world, float deltaTime)
    {
        var gravity = entity.GetRequiredComponent<AffectedByGravity>().Gravity;
        var position = entity.GetRequiredComponent<Position>();
        var size = entity.GetComponent<Size>() ?? new Size(1, 1);

        var originalPosition = new Position(position.Left, position.Top);

        position.Top += gravity * deltaTime;

        if (entity.HasComponent<Velocity>())
        {
            HandleVelocity(entity.GetRequiredComponent<Velocity>(), position, deltaTime);
        }

        if (entity.HasComponent<LimitedByBounds>())
        {
            LimitBounds(position, size);
            LimitBySolidEntities(world, size, position, originalPosition);
        }
    }

    private static void LimitBounds(Position position, Size size)
    {
        if (position.Left < 0)
        {
            position.Left = 0;
        }
        if (position.Left + size.Width >= UiInterface.WorldWidth)
        {
            position.Left = UiInterface.WorldWidth - size.Width;
        }
        if (position.Top < UiInterface.WorldTop)
        {
            position.Top = UiInterface.WorldTop;
        }
        if (position.Top + size.Height >= UiInterface.WorldBottom)
        {
            position.Top = UiInterface.WorldBottom;
        }
    }

    private static void HandleVelocity(Velocity velocity, Position position, float deltaTime)
    {
        position.Top -= velocity.Y * deltaTime;
        position.Left += velocity.X * deltaTime;

        velocity.X *= 1.0f - deltaTime * Configuration.MovementDecayRate;
        velocity.Y *= 1.0f - deltaTime * Configuration.MovementDecayRate;

        // Zero out very small values to prevent jitter
        if (Math.Abs(velocity.X) < 0.1) velocity.X = 0;
        if (Math.Abs(velocity.Y) < 0.1) velocity.Y = 0;
    }

    private static void LimitBySolidEntities(World world, Size size, Position position, Position originalPosition)
    {
        if (IsBlocked(position, size, world.GetEntitiesWith(typeof(Position), typeof(Solid))))
        {
            position.Top = originalPosition.Top;
            if (IsBlocked(position, size, world.GetEntitiesWith(typeof(Position), typeof(Solid))))
            {
                position.Left = originalPosition.Left;
            }
        }
    }

    private static bool IsBlocked(Position position, Size size, IEnumerable<Entity> solidPositions)
    {
        return solidPositions.Any(s =>
            s.GetRequiredComponent<Position>().LeftInt < position.LeftInt + size.Width &&
            s.GetRequiredComponent<Position>().LeftInt + s.GetRequiredComponent<Size>().Width > position.LeftInt &&
            s.GetRequiredComponent<Position>().TopInt < position.TopInt + size.Height &&
            s.GetRequiredComponent<Position>().TopInt + s.GetRequiredComponent<Size>().Height > position.TopInt);
    }
}

File: Web/src/systems/PhysicsSystem.js
import ISystem from './ISystem.js';
import UiInterface from '../core/uiInterface.js';
import { Velocity, Position, AffectedByGravity, Size, Solid, LimitedByBounds } from '../components/Components.js';
import Configuration from '../core/configuration.js';

class PhysicsSystem extends ISystem {
    update(world, deltaTime) {
        world.getEntitiesWith(Position, AffectedByGravity).forEach(e =>
            this.updateEntity(e, world, deltaTime)
        );
    }

    updateEntity(entity, world, deltaTime) {
        const gravity = entity.getRequiredComponent(AffectedByGravity).Gravity;
        const position = entity.getRequiredComponent(Position);
        const size = entity.getComponent(Size) || new Size(1, 1);

        const originalPosition = new Position(position.Left, position.Top);


        if (entity.hasComponent(Velocity)) {
            const velocity = entity.getRequiredComponent(Velocity);
            velocity.Y -= gravity * deltaTime;
            this.handleVelocity(velocity, position, deltaTime);
        }

        if (entity.hasComponent(LimitedByBounds)) {
            const velocity = entity.getComponent(Velocity);
            this.limitBounds(position, size, velocity);
            this.limitBySolidEntities(world, size, position, originalPosition, velocity);
        }
    }

    limitBounds(position, size, velocity) {
        if (position.Left < 0) {
            position.Left = 0;
        }
        if (position.Left + size.Width >= UiInterface.WorldWidth) {
            position.Left = UiInterface.WorldWidth - size.Width;
        }
        if (position.Top < UiInterface.WorldTop) {
            position.Top = UiInterface.WorldTop;
        }
        if (position.Top + size.Height >= UiInterface.WorldBottom) {
            position.Top = UiInterface.WorldBottom;
            velocity.Y = 0;
        }
    }

    handleVelocity(velocity, position, deltaTime) {
        position.Top -= velocity.Y * deltaTime;
        position.Left += velocity.X * deltaTime;

        velocity.X *= 1.0 - deltaTime * Configuration.MovementDecayRate;

        // Zero out very small values to prevent jitter
        if (Math.abs(velocity.X) < 0.1) {
            velocity.X = 0;
        }
        else {
            position.Direction = velocity.X < 0 ? 'l' : 'r';
        }
        if (Math.abs(velocity.Y) < 0.1) velocity.Y = 0;

    }

    limitBySolidEntities(world, size, position, originalPosition, velocity) {
        if (this.isBlocked(position, size, world.getEntitiesWith(Position, Solid))) {
            position.Top = originalPosition.Top;
            velocity.Y = 0;

            if (this.isBlocked(position, size, world.getEntitiesWith(Position, Solid))) {
                position.Left = originalPosition.Left;
            }
        }
    }

    isBlocked(position, size, solidPositions) {
        return solidPositions.some(s => {
            const sPos = s.getRequiredComponent(Position);
            const sSize = s.getRequiredComponent(Size);
            return sPos.LeftInt < position.LeftInt + size.Width &&
                sPos.LeftInt + sSize.Width > position.LeftInt &&
                sPos.TopInt < position.TopInt + size.Height &&
                sPos.TopInt + sSize.Height > position.TopInt;
        });
    }
}

export default PhysicsSystem;